Implement a searchable and expandable configuration property table

FroMage opened the pull request:
This builds on https://github.com/quarkusio/quarkus/pull/4050 to make the config key table collapsible using JS. I think it should work out on PDF too since most of the behaviour is controlled by JS/CSS.

Please @gsmet @emmanuelbernard @machi1990 try it on for size.

Collapsed:

![Screenshot from 2019-10-04 18-20-07](https://user-images.githubusercontent.com/179265/66223477-e1a9cb80-e6d3-11e9-97fd-ab18e5e8c2be.png)

Expanded:

![Screenshot from 2019-10-04 18-20-16](https://user-images.githubusercontent.com/179265/66223471-dfe00800-e6d3-11e9-9013-75449d1de5a3.png)

I'm not entirely sure of the behaviour of expanding the cell to the whole table, but it felt better to read the expanded docs. It has the drawback of hiding the default value and type, but I guess I could dynamically add it when expanding if you think it's important.

I have added collapse/expand all links on each collapsible row, rather than at the top of the table because I'm thinking of the "all config" table and assume that link will be out of reach. I could add it _as well_ though if you think it's important.

I think we can remove the unstructured view with this, because it's just a lot more readable (don't look at the font size, I wish I knew why it became so small all of a sudden, I intend to fix it).

For the "but I can't search collapsed text" I hope users will click on "expand all". Or I can add a search box on every table which will expand all rows that match the searched text. I believe that would be very nice.

Anyway, the pictures don't do it justice to feel the UX: please check it out, compile and open `docs/target/generated-docs/hibernate-orm-guide.html` once it's all compiled/rendered.

machi1990 committed:
docs: add support of config doc sections in generated docs

docs: add a possibility to specify a key of a map config item through @ConfigDocMapKey

Fixes #4045

FroMage committed:
Implement a searchable/expandable config table

Also remove the detailed view as it's not useful anymore.

FroMage commented:
I just added interactive search, this is really starting to be awesome!

emmanuelbernard commented:
> For the "but I can't search collapsed text" I hope users will click on "expand all". Or I can add a search box on every table which will expand all rows that match the searched text. I believe that would be very nice.

To me the problem is more for the user landing on the guide and CTRL+F not knowing there are tables with collapsable content.

FroMage commented:
It is, but try it.

For the "all config properties" page it's not much of an issue because the search box will be on top.

For the other guides there's a good chance Ctrl-F will land them on the table anyway and they will be able to see the search box.

But again, look at it, try it. My opinion is that it's an acceptable compromise because it's much better than the unstructured layout and once you learn to use the search box instead of Ctrl-F it's pretty amazing. But frankly this is subjective, you have to try it to decide if the benefits outweight the risk.

gsmet commented:
Having a search feature is really nice.

I would keep the 3 columns even in expanded state. If you expand all, your are missing 2 very important information.

I see you also have removed the legend. I think it has some value to have it as this notion is not obvious for people coming from other frameworks.

I would be interested in you testing the Expand buttons with the long keys of this guide: file:///home/gsmet/git/quarkus-stable/docs/target/generated-docs/hibernate-search-guide.html to be sure it's OK.

We also need to check the layout on the website. The font is significantly larger there.

Another point is that some config sections had intros. IIRC, @gwenneg 's PR had some. I don't know if it fits in this model and if we could add the intro below the config section title (the one of the table) somehow.

FroMage commented:
> I would keep the 3 columns even in expanded state. If you expand all, your are missing 2 very important information.

I can, but then the text is hard to read on a single column. I can add the two important info to the expanded cell, though.

> I see you also have removed the legend

I did not manage to add the legend to the table caption. Asciidoc kicked it out. I figured the on-hover label could be enough. I can try to re-add it.

> I would be interested in you testing the Expand buttons with the long keys of this guide: file:///home/gsmet/git/quarkus-stable/docs/target/generated-docs/hibernate-search-guide.html to be sure it's OK.

Will try.

> We also need to check the layout on the website. The font is significantly larger there.

Yeah, I want to make the font the same size as the rest of the text.

> Another point is that some config sections had intros. IIRC, @gwenneg 's PR had some. I don't know if it fits in this model and if we could add the intro below the config section title (the one of the table) somehow.

Ah, good point, lemme try something.

FroMage commented:
- Added the legend back
- Fixed table font size to the default font size for asciidoc tables
- Added the type/default info in the cell when expanding
- Checked the long keys: yes they make the config key line break, but long lines break anyway, so I think it's acceptable and rare

FroMage commented:
I think the most pressing remaining issue is the "all the config" page now.

emmanuelbernard commented on core/runtime/src/main/java/io/quarkus/runtime/annotations/ConfigItem.java:
```diff
@@ -53,4 +53,11 @@
     String defaultValue() default NO_DEFAULT;
+
+    /**
+     * Specify the default value documentation, when it cannot be expressed by @{link #defaultValue()}.
```
A small example like the Hibernate one could help usability I think. But otherwise that looks good.

FroMage replied:
Done.

machi1990 commented:
> I think we can remove the unstructured view with this, 

Hey @FroMage I think so too. 

Also +1 for having a search box. 

I have not looked how a config section section with a bit of introduction texts will look like, I’ll try to add some texts and see.

FroMage commented:
@machi1990 shall I merge it then? It contains your #3803 changes.

FroMage commented:
> I have not looked how a config section section with a bit of introduction texts will look like, I’ll try to add some texts and see.

I didn't add support for this yet, because I didn't find docs which had them. Do we have any? I think we can add it later.

machi1990 commented:
> @machi1990 shall I merge it then? It contains your #3803 changes.

I added one commit on the PR, maybe if we cherry-pick it here I can close my PR. I also think that @gsmet wanted to review the code and also add introductory notes for hibernate config sections.

gsmet commented:
@FroMage have you removed the detailed view in your latest PR?

machi1990 commented:
> > @machi1990 shall I merge it then? It contains your #3803 changes.
> 
> I added one commit on the PR, maybe if we cherry-pick it here I can close my PR. 

I was referring to this commit - https://github.com/quarkusio/quarkus/pull/4050/commits/f3b364f119dbe9bc97b8b12367f1c11facaaf5eb which apart from adding a link to config sections, refactored the parsing config section so that it retrieves the title and section details in one pass.

FroMage commented:
OK, merged your commit, removed StringBuilder in favour of Writer (should be much faster), removed unstructured section.

gsmet commented:
@FroMage have you checked how it looks in the website?

Last time I pushed significant changes, I had to adjust things a bit in the code when I tried to include that in the website.

FroMage committed:
First draft of page displaying all the configuration properties

FroMage commented:
I've removed the links to the unstructured view which is gone, and I added an `all-config` page :)

It's a bit slow when filtering (it's a huge table) so I'd like to work on improving that before we announce the page or link to it, but I figured it's still nice for us to discuss the page since we know its "secret" url.

Also it's missing extension names so it's not Good Enough™ ATM.

gsmet reviewed (approved):
Stéphane made some adjustments for the website. Now we should be OK.

machi1990 commented:
Nice work @FroMage @gsmet 👍

gsmet commented:
I haven't done anything except complaining :).
